spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala

```diff
@@ -103,6 +108,25 @@ class DeltaHistoryManager(
     monotonizeCommitTimestamps(info.collect()).reverse
   }
 
+  /**
+   * Returns the latest commit that happened at or before `time` in the range `[start, end)`.
+   * All the commits in the range `[start, end)` are assumed to not have inCommitTimestamps.
+   * If no such commit exists, the earliest commit is returned.
+   */
+  def getCommitFromNonICTRange(start: Long, end: Long, time: Long): Commit = {
+    if (end - start > 2 * maxKeysPerList) {
+      parallelSearch(time, start, end)
+    } else {
+      val commits = getCommits(
+        deltaLog.store,
+        deltaLog.logPath,
+        start,
+        Some(end),
+        deltaLog.newDeltaHadoopConf())
+      lastCommitBeforeTimestamp(commits, time).getOrElse(commits.head)
```

Comment:
I believe this is only possible in extremely rare scenarios and even in those cases, it will end up throwing some unrecoverable error downstream. This part of the code is mostly untouched: https://github.com/delta-io/delta/pull/2813/files#diff-98df89d4b0ce76abb5263d4fbb6f991660083a254dcdb9d922a4973a9937a4e2L139